src/main.cpp

```diff
@@ -3018,6 +3019,8 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
                     state = CValidationState();
                     fInvalidFound = true;
                     fContinue = false;
+                    // If we didn't actually connect the block, don't notify listeners about it
+                    connectTrace.blocksConnected.pop_back();
```

Comment:
nit: slightly weird that the caller of ConnectTip() has to know to do this on failures.  Perhaps document this somewhere?